packages/web-components/src/drawer/drawer.ts

```diff
@@ -0,0 +1,244 @@
+import { attr, FASTElement, observable, Updates } from '@microsoft/fast-element';
+import { eventAnimationEnd, keyEscape } from '@microsoft/fast-web-utilities';
+import { DrawerModalType, DrawerPosition, DrawerSize, DrawerType } from './drawer.options.js';
+
+export class Drawer extends FASTElement {
+  private closing: boolean = false;
+
+  public connectedCallback(): void {
+    super.connectedCallback();
+
+    Updates.enqueue(() => {
+      this.setComponent();
+    });
+  }
+
+  public disconnectedCallback(): void {
+    super.disconnectedCallback();
+  }
+
+  /**
+   * The dialog element.
+   * @public
+   *
+   */
+  @observable
+  public dialog!: HTMLDialogElement;
+
+  /**
+   * Determines whether the drawer should be displayed as modal, non-modal, or alert.
+   * When in modal or alert mode, an overlay is applied over the rest of the view.
+   * @public
+   * @remarks
+   * HTML Attribute: modalType
+   */
+  @attr({ attribute: 'modal-type' })
+  public modalType: DrawerModalType = DrawerModalType.modal;
+
+  /**
+   * The ID of the element that labels the drawer.
+   * @public
+   * @remarks
+   * HTML Attribute: aria-labelledby
+   */
+  @attr({ attribute: 'aria-labelledby' })
+  public ariaLabelledby?: string;
+
+  /**
+   * The ID of the element that describes the drawer.
+   * @public
+   * @remarks
+   * HTML Attribute: aria-describedby
+   */
+  @attr({ attribute: 'aria-describedby' })
+  public ariaDescribedby?: string;
+
+  /**
+   * Sets the type of the drawer (overlay/inline).
+   * @public
+   * @remarks
+   * HTML Attribute: type
+   * @defaultValue overlay
+   */
+  @attr
+  public type?: DrawerType;
+
+  /**
+   * Sets the position of the drawer (start/end).
+   * @public
+   * @remarks
+   * HTML Attribute: position
+   * @defaultValue start
+   */
+  @attr
+  public position?: DrawerPosition;
+
+  /**
+   * Sets the size of the drawer (small/medium/large).
+   * @public
+   * @remarks
+   * HTML Attribute: size
+   * @defaultValue medium
+   */
+  @attr({ attribute: 'size' })
+  public size?: DrawerSize;
+
+  /**
+   * Sets the open state of the drawer
+   * @public
+   * @remarks
+   * HTML Attribute: open
+   * @defaultValue false
+   */
+  @attr({ mode: 'boolean' })
+  public open: boolean = false;
+
+  /**
+   * Emits a 'cancel' event.
+   * @internal
+   */
+  public cancel(): void {
+    this.$emit('cancel');
+  }
+
+  /**
+   * Handles changes to the `open` property.
+   * @public
+   */
+  public openChanged(oldValue: boolean, newValue: boolean): void {
+    if (newValue !== oldValue) {
```

Comment:
This is perhaps a generic infra question but does the infra actually allow this openChanged callback ever to be called when newValue === oldVaue (ie when open is *not* changed)? That would seem a bit semantically improper.
If not then the if (newValue !== oldValue) condition would not be needed at the top of any attr change handlers.
Either way (not required or required and satisfied) once newValue !== oldValue is satisfied I don't think you need to check both values in the subsequent conditionals and they should simplify to
if (newValue) {
this.show();
} else {
 this.hide();
}

Reply:
You're correct. In a well-designed system, an *Changed callback should only be invoked when the value has actually changed, i.e., newValue !== oldValue. Therefore, the check newValue !== oldValue might be redundant.

However, sometimes these checks are added as a defensive programming measure, to prevent unexpected behavior in case of any bugs or inconsistencies in the system.

Reply:
> This is perhaps a generic infra question but does the infra actually allow this openChanged callback ever to be called when newValue === oldVaue (ie when open is _not_ changed)? That would seem a bit semantically improper. If not then the if (newValue !== oldValue) condition would not be needed at the top of any attr change handlers. Either way (not required or required and satisfied) once newValue !== oldValue is satisfied I don't think you need to check both values in the subsequent conditionals and they should simplify to if (newValue) { this.show(); } else { this.hide(); }

If the value doesn't change, the callback should not be called. If it's being called, it's safe to assume it has been changed. 

Code comes at a cost, we should be defensive _only_ in scenarios where necessary.